File: bindings/swift/examples/TransactionSignerCallback.swift
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import Foundation
import IotaSDK

final class AsyncSigner: TransactionSignerFn, @unchecked Sendable {
  let key: Ed25519PrivateKey

  init(key: Ed25519PrivateKey) {
    self.key = key
  }

  func sign(transaction: Transaction) async throws -> TransactionSignerFnOutput {
    return TransactionSignerFnOutput(signature: try key.signTransaction(transaction: transaction))
  }
}

@main
struct TransactionSignerCallbackExample {
  static func main() async throws {
    let amount: UInt64 = 1000
    let recipientAddress = try Address.fromHex(
      hex: "0x0000a4984bd495d4346fa208ddff4f5d5e5ad48c21dec631ddebc99809f16900")

    let privateKey = try Ed25519PrivateKey(bytes: Data(repeating: 0, count: 32))
    let publicKey = privateKey.publicKey()
    let senderAddress = publicKey.deriveAddress()
    print("Sender address: \(senderAddress.toHex())")

    let client = GraphQlClient.newLocalnet()

    // Request funds from faucet
    let faucet = FaucetClient.newLocalnet()
    _ = try await faucet.requestAndWaitForFinalized(address: senderAddress, client: client)

    let builder = client.transactionBuilder(sender: senderAddress)
    _ = builder.sendIota(recipient: recipientAddress, amount: PtbArgument.u64(value: amount))

    let signer = TransactionSigner(signerFn: AsyncSigner(key: privateKey))
    let effects = try await builder.execute(signer: signer, waitFor: WaitForTx.finalized)

    print("Digest: \(hexEncode(input: effects.digest().toBytes()))")
    print("Transaction status: \(effects.asV1().status)")
    print("Effects: \(effects.asV1())")
  }
}

File: bindings/python/examples/transaction_signer_callback.py
# Copyright (c) 2025 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

from lib.iota_sdk import *

import asyncio


class AsyncSigner(TransactionSignerFn):

    def __init__(self, key: Ed25519PrivateKey):
        self.key = key

    async def sign(self, transaction: Transaction):
        return TransactionSignerFnOutput(
            signature=self.key.sign_transaction(transaction))


async def main():
    amount = 1000
    recipient_address = Address.from_hex(
        "0x0000a4984bd495d4346fa208ddff4f5d5e5ad48c21dec631ddebc99809f16900")

    private_key = Ed25519PrivateKey(b"\x00" * 32)
    public_key = private_key.public_key()
    sender_address = public_key.derive_address()
    print(f"Sender address: {sender_address.to_hex()}")

    client = GraphQlClient.new_localnet()

    # Request funds from faucet
    faucet = FaucetClient.new_localnet()
    await faucet.request_and_wait_for_finalized(sender_address, client)

    builder = client.transaction_builder(sender_address)
    builder.send_iota(recipient_address, PtbArgument.u64(amount))

    signer = TransactionSigner(AsyncSigner(private_key))
    effects = await builder.execute(signer, WaitForTx.FINALIZED)

    print(f"Digest: {hex_encode(effects.digest().to_bytes())}")
    print(f"Transaction status: {effects.as_v1().status()}")
    print(f"Effects: {effects.as_v1()}")


if __name__ == "__main__":
    asyncio.run(main())
